### DISCUSSION TITLE
Assertion API

### DISCUSSION THREAD

**Discussion by schoeberl** [chiselverify/chiselverify#26] [Ideas]
ScalaTest has the following notion on _eventually_ (from the Scala book):

```
val xs = 1 to 3
val it = xs.iterator
eventually { it.next() shouldBe 3 }
```

Maybe we should follow this style for our timed assertions.

**Comment by dobios** [comment]
Yes I agree, that is a lot more elegant! You can see PR #28 for a potential solution.

**Reply by dobios** [reply]
@schoeberl So, if I understood your proposal, we should then have the following syntax for timed assertions:
```scala
eventually { a == b } delay by 3
always { a > b } delay by 5 
exactly { a < b } delay by 7
```  
Is it something like that you are looking for?

**Reply by dobios** [reply]
@schoeberl For the operators, I can't use any syntax that potentially collides with chisel syntax, so I chose to have them preceded by the `?` character:  
```scala 
eventually { a ?== b } delay by 2
always { a ?< b } delay by 5
exactly { a ?> b } delay by 7
never { a ?>= b } delay by 1
```
It does look a bit odd, so I'm always open to any other proposal in this regard.

**Comment by schoeberl** [comment]
Mmh, I am not sure if I like the preceding '?'. Do we really need this? The expression is just a function returning a Scala Boolean, right? So we can use any ChiselTest code (peek). I don't understand why this collides with Chisel operators, we are in Scala land, right?

Another proposal: for quick and dirty tests we can have an optional delay (say 100 clock cycles) to state:

```
eventually {
  dut.io.a.peek().litVal == 0
} 
```
Maybe having the delay as a parameter, such as:
```
eventually(20) { ...}

**Comment by schoeberl** [comment]
I think the proposed syntactic sugar is orthogonal to the timed assertion topic. It could also be used in general testing. I would stick to _normal_ Chisel test code within the code block. We shall discuss if we want to add those funny operators to ChiselVerify in general. However, there is a complexity danger here when adding too many operators. We might end up beating SystemVerilog by having more than 250 keywords/operators in Chisel/ChiselTest/ChiselVerify.

Let's move the operator discussion to https://github.com/chiselverify/chiselverify/discussions/29

**Reply by dobios** [reply]
The idea behind the operators is to add the possibility to compare two different ports at different cycles, which isn't really as easy to do using the regular conditions.  For example:  
```scala  
eventually(5) { dut.io.a ?== dut.io.b }
```
Would have to be done in a much more verbose way otherwise:  
```scala
eventually(5) { () => { 
    val a = dut.io.a.peek().litValue()
    dut.clock.step(5)
    a == dut.io.b.peek().litValue()
}}
```
I personally find that the timed operators remove a lot of headache and simplify the overall syntax.

**Comment by schoeberl** [comment]
Mmh, shouldn't eventually be a condition that is eventually true? I did not think about a comparison of values from different clock cycles. This might be interesting as well, but a different use case. If we mimic ScalaTest, then it is a boolean condition that might become true some cycle within the limit. We should look at what SV has for a double-check.

```eventually``` should just consume a function literal that returns a Boolean. The ```step()``` and ```assert()``` is within the ```eventually``` function.  So your example should look like this:

```
eventually() {
  dut.io.a.peek().litValue() == dut.io.b.peek().litValue()
}
```

BTW: did I mention that I think ```litValue``` is *very* misleading when using it in a testing context.

**Reply by dobios** [reply]
Yes that works if you just want to evaluate your function at multiple different cycles, but not if you want to do something like compare two ports at different cycles (i.e. port `a` from cycle 0 compared to port `b` from cycle 7). To do this you need some sort of separation between the two ports, which is offered by the operators. Such a separation isn't possible to get by just using a function literal.    
  
Otherwise the stepping and asserting is already done internally, but the assertion simply evaluates the given function: `assert(func())`.

**Reply by schoeberl** [reply]
Yes, I think this is what I would want for eventually. Same as in ScalaTest. Comparing values from different points in time is a very special case. And you can do it like this:

```
val a = dut.io.a.peek().litValue()
eventually(5) {
  dut.io.b.peek().litValue() == a
}
```

**Comment by schoeberl** [comment]
We shall use by-name paramters to drop ```() => ```

```
  def expect(cycles: Int)(f: => Boolean): Unit = {
    println(cycles)
    println(f)
  }

  def expect (f: => Boolean): Unit = expect(150)(f)


  expect(5) {
    val a = 1
    val b = 2
    a == b
  }

  expect {
    true
  }
```
